### ISSUE TITLE
Get `and`, `or`, `not`, `implies` to work directly on `Prop`

### ISSUE THREAD

**Issue by byorgey** [disco-lang/disco#316]
Right now operators like `and`, `or`, `not`, `implies` must only be used on `Bool`.  This means we can't have things like `(forall n : N. blah) /\ (forall y : Integer. foo)` even though it seems like that should be a perfectly reasonable proposition.

**Comment by byorgey** [comment]
I think this would require a bunch of work to make `:test` work.  I am not quite sure what exactly would be required.

**Comment by byorgey** [comment]
Hmm, it looks like this is already kind of supported, see the `prop-higher-order` test!  So maybe this can be done just as syntax sugar?

**Comment by byorgey** [comment]
For reference, `test/prop-higher-order/higher-order.disco` defines
```
pand : Prop * Prop -> Prop
pand(p, q) = forall and_side : Unit + Unit. {?
    p when and_side is left _,
    q otherwise
  ?}

all : List(Prop) -> Prop
all ps = reduce(pand, true, ps)

por : Prop * Prop -> Prop
por(p, q) = exists or_side : Unit + Unit. {?
    p when or_side is left _,
    q otherwise
  ?}

any : List(Prop) -> Prop
any ps = reduce(por, false, ps)
```

It's cool to know that this can indeed be done given only what we have currently, but I think things would benefit a lot from built-in support for `/\`, `\/` etc. on `Prop`.  For example, test result messages for `pand` and `por` mention `and_side` and `or_side`:

```
Disco> :test pand(forall x: Bool. x == x, forall y: Bool. y == not y)
  - Test is false: pand(∀x. x == x, ∀y. y == not y)
    Counterexample:
      and_side = right(■)
      y        = false
```
It would be much better to have something like
```
  - Test is false: (∀x. x == x) /\ (∀y. y == not y)
    - Because (∀y. y == not y) is false
      Counterexample:
        y        = false
```
Of course, it's quite possible that we could desugar `/\`, `\/` as in `pand` and `por`, but retain the knowledge of how they were built so that we can display better messages/explanations.

**Comment by byorgey** [comment]
Hah, I just found this comment in the source that I had forgotten about:

https://github.com/disco-lang/disco/blob/ca63dc8db061e7bd781de4c5f99cafe06372527a/src/Disco/Typecheck.hs#L570-L580

See https://github.com/disco-lang/disco/commit/9c49be81c84a390e76b23dffb4fc0645a4164b9c .

**Comment by byorgey** [comment]
The next step would be to desugar operators like `/\`, `\/`, etc. into special `Prop`-formers in the `Core` language; then interpret them appropriately.

**Comment by byorgey** [comment]
Closed by #365 .
